src/dotty/tools/dotc/transform/TreeChecker.scala

```diff
@@ -181,6 +181,106 @@ class TreeChecker extends Phase with SymTransformer {
+      // case tree: untpd.Literal =>
+      // case tree: untpd.New =>
+      case Pair(left, right) =>
+        assertIdentNotJavaClass(left)
+        assertIdentNotJavaClass(right)
+      case Typed(expr, _) =>
+        assertIdentNotJavaClass(expr)
+      case NamedArg(_, arg) =>
+        assertIdentNotJavaClass(arg)
+      case Assign(_, rhs) =>
+        assertIdentNotJavaClass(rhs)
+      case Block(stats, expr) =>
+        stats.foreach(assertIdentNotJavaClass _)
+        assertIdentNotJavaClass(expr)
+      case If(_, thenp, elsep) =>
+        assertIdentNotJavaClass(thenp)
+        assertIdentNotJavaClass(elsep)
+      case untpd.Function(_, body) =>
+        assertIdentNotJavaClass(body)
+      // case tree: untpd.Closure =>
+      case Match(selector, cases) =>
+        assertIdentNotJavaClass(selector)
+        cases.foreach(caseDef => assertIdentNotJavaClass(caseDef.body))
+      case Return(expr, _) =>
+        assertIdentNotJavaClass(expr)
+      case Try(expr, cases, finalizer) =>
+        assertIdentNotJavaClass(expr)
+        cases.foreach(caseDef => assertIdentNotJavaClass(caseDef.body))
+        assertIdentNotJavaClass(finalizer)
+      case untpd.Throw(expr) =>
+        assertIdentNotJavaClass(expr)
+      // case tree: untpd.TypeApply =>
+      // case tree: untpd.Super =>
+      case SeqLiteral(elems, _) =>
+        elems.foreach(assertIdentNotJavaClass)
+      // case tree: untpd.TypeTree =>
+      // case tree: untpd.SingletonTypeTree =>
+      // case tree: untpd.AndTypeTree =>
+      // case tree: untpd.OrTypeTree =>
+      // case tree: untpd.RefinedTypeTree =>
+      // case tree: untpd.AppliedTypeTree =>
+      // case tree: untpd.ByNameTypeTree =>
+      // case tree: untpd.TypeBoundsTree =>
+      // case Alternative(trees) =>
+      // case tree: untpd.PackageDef =>
+      case Annotated(_, arg) =>
+        assertIdentNotJavaClass(arg)
+      // case tree: untpd.TypedSplice =>
+      // case tree:  untpd.UnApply =>
+      case untpd.PostfixOp(od, _) =>
+        assertIdentNotJavaClass(od)
+      case untpd.InfixOp(left, _, right) =>
+        assertIdentNotJavaClass(left)
+        assertIdentNotJavaClass(right)
+      case untpd.PrefixOp(_, od) =>
+        assertIdentNotJavaClass(od)
+      case untpd.Tuple(trees) =>
+        trees.foreach(assertIdentNotJavaClass)
+      case untpd.WhileDo(_, body) =>
+        assertIdentNotJavaClass(body)
+      case untpd.DoWhile(body, _) =>
+        assertIdentNotJavaClass(body)
+      case untpd.ForYield(_, expr) =>
+        assertIdentNotJavaClass(expr)
+      case untpd.ForDo(_, body) =>
+        assertIdentNotJavaClass(body)
+      // case untpd.GenFrom(_, expr) =>
+      case untpd.GenAlias(_, expr) =>
+        assertIdentNotJavaClass(expr)
+      // case untpd.ContextBounds(_, _) =>
+      case _ =>
+    }
+
+    override def typed(tree: untpd.Tree, pt: Type = WildcardType)(implicit ctx: Context): tpd.Tree = {
+      checkIdentNotJavaClass(tree)
```

Comment:
Thanks, I'll follow the approach.

BTW, it's safe to assume all untyped ASTs are desugared after `FrontEnd`?